File: code/packages/go/interrupt-handler/frame.go
package interrupthandler

// =========================================================================
// Interrupt Frame (Saved CPU Context)
// =========================================================================

// InterruptFrame holds all CPU state needed to resume execution after an
// interrupt has been handled. When an interrupt fires, the CPU pushes this
// frame onto the kernel stack before jumping to the ISR. When the ISR
// returns, the CPU pops the frame and resumes the interrupted code.
//
// Layout on the kernel stack (136 bytes total):
//
//	+---------------------------------------+ <- old sp
//	| PC (return address)         | 4 bytes |
//	| MStatus register            | 4 bytes |
//	| MCause register             | 4 bytes |
//	| x1  (ra - return address)   | 4 bytes |
//	| x2  (sp - stack pointer)    | 4 bytes |
//	| x3  (gp - global pointer)   | 4 bytes |
//	| ...                         |         |
//	| x31 (t6)                    | 4 bytes |
//	+---------------------------------------+ <- new sp
//	  Total: 3 + 31 = 34 words = 136 bytes
//
// Why save ALL 32 registers? The ISR is arbitrary code -- it might use any
// register. If we only saved "some" registers, the ISR could corrupt the
// interrupted program's state. Saving everything is safe and simple.
//
// Real CPUs optimize this: the ISR saves only the registers it actually
// uses (callee-saved convention). We save all 32 for correctness.
type InterruptFrame struct {
	PC        uint32     // Saved program counter (where to resume)
	Registers [32]uint32 // All 32 RISC-V general-purpose registers (x0-x31)
	MStatus   uint32     // Machine status register
	MCause    uint32     // What caused the interrupt (interrupt number)
}

// SaveContext creates an InterruptFrame from the current CPU state.
// This is called at the beginning of interrupt handling, before the ISR
// runs. The mcause field records which interrupt triggered the save.
//
// Parameters:
//   - registers: all 32 general-purpose registers (x0-x31)
//   - pc: the program counter (address of the next instruction to execute
//     after the interrupt is handled)
//   - mstatus: the machine status register
//   - mcause: the interrupt number that caused this context save
//
// Returns a complete InterruptFrame that can later be passed to
// RestoreContext to resume the interrupted code.
func SaveContext(registers [32]uint32, pc uint32, mstatus uint32, mcause uint32) InterruptFrame {
	return InterruptFrame{
		PC:        pc,
		Registers: registers,
		MStatus:   mstatus,
		MCause:    mcause,
	}
}

// RestoreContext extracts CPU state from an InterruptFrame. This is called
// after the ISR completes, to resume the interrupted code exactly where it
// left off.
//
// Returns:
//   - registers: all 32 general-purpose registers
//   - pc: the program counter to resume at
//   - mstatus: the machine status register
func RestoreContext(frame InterruptFrame) (registers [32]uint32, pc uint32, mstatus uint32) {
	return frame.Registers, frame.PC, frame.MStatus
}

File: code/packages/rust/interrupt-handler/src/frame.rs
//! Interrupt Frame — saved CPU context for interrupt entry/exit.
//!
//! When an interrupt fires, the CPU saves everything needed to resume the
//! interrupted code later. This is the interrupt frame (or trap frame).
//!
//! # Layout (136 bytes)
//!
//! ```text
//! PC (return address)         4 bytes
//! MStatus register            4 bytes
//! MCause register             4 bytes
//! x1-x31 (31 registers)      124 bytes
//! Total: 34 words = 136 bytes
//! ```
//!
//! Why save ALL 32 registers? The ISR is arbitrary code -- it might use any
//! register. Saving everything is safe and simple.

/// Holds all CPU state needed to resume after an interrupt.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct InterruptFrame {
    /// Saved program counter (where to resume).
    pub pc: u32,
    /// All 32 RISC-V general-purpose registers (x0-x31).
    pub registers: [u32; 32],
    /// Machine status register.
    pub mstatus: u32,
    /// What caused the interrupt (interrupt number).
    pub mcause: u32,
}

impl Default for InterruptFrame {
    fn default() -> Self {
        Self {
            pc: 0,
            registers: [0; 32],
            mstatus: 0,
            mcause: 0,
        }
    }
}

/// Create an InterruptFrame from the current CPU state.
///
/// Called at the beginning of interrupt handling, before the ISR runs.
/// The registers are copied (not referenced) for safety.
pub fn save_context(registers: [u32; 32], pc: u32, mstatus: u32, mcause: u32) -> InterruptFrame {
    InterruptFrame {
        pc,
        registers,
        mstatus,
        mcause,
    }
}

/// Extract CPU state from an InterruptFrame.
///
/// Called after the ISR completes, to resume the interrupted code.
/// Returns `(registers, pc, mstatus)`.
pub fn restore_context(frame: &InterruptFrame) -> ([u32; 32], u32, u32) {
    (frame.registers, frame.pc, frame.mstatus)
}
